### PR TITLE
Update vsock logic to address stalling problems

### PR DESCRIPTION
**Pull request by gc-plp**
## Reason for This PR

Addresses #1751 

## Description of Changes

This PR changes some of the internal mechanisms in the vsock device, that were causing the problems described in #1751.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

**Comment by sandreim** [comment]
Is the issue fixed by this PR ?

**Comment by gc-plp** [comment]
> Is the issue fixed by this PR ?

Hi Andrei. I added more details in the comment [here](https://github.com/firecracker-microvm/firecracker/issues/1751#issuecomment-630281988)

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 1]
```diff
@@ -143,16 +143,22 @@ impl VsockChannel for VsockMuxer {
...
                         .set_fwd_cnt(0);
+                    self.rxq.pop().unwrap();
                     return Ok(());
                 }
 
                 // We'll defer building the packet to this connection, since it has something
                 // to say.
                 MuxerRx::ConnRx(key) => {
                     let mut conn_res = Err(VsockError::NoData);
+                    let mut do_pop = true;
```
[the remark this answers is not in the document]
Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs:154]: > it is necessary to keep the connection inside the muxer rxq after handling a pending RX instruction for that specific connection

The purpose of this change is to keep the connection inside the `rxq`, in the situation where there are multiple flags pending.

Previously to this, the muxer would permanently pop a connection from the `rxq` even if it had multiple flags pending, which made the unhandled flag to be dropped and lost forever. If that flag would have been a credit update, the flow control would be broken.

I agree that instructions for one connection would be prioritized, but since the instructions are flag based and multiple similar instructions cannot be queued at the same time, I don't think that there is a significant impact on neighboring connections. 
We could implement a _fair_ scheduler, but we also need to keep in mind that this code path is executed very often and we need an efficient way to look-up flags and pop elements.

If you have a proposal on this, let's discuss it.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 2]
```diff
@@ -439,7 +452,13 @@ where
                         "vsock: error flushing TX buf for (lp={}, pp={}): {:?}",
                         self.local_port, self.peer_port, err
                     );
-                    self.kill();
+                    match err {
+                        Error::TxBufFlush(inner) if inner.kind() == ErrorKind::WouldBlock => {
+                            // This should never happen (EWOULDBLOCK after EPOLLOUT), but
```
[the remark this answers is not in the document]
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs:457]: Yes, you can get `WouldBlock` errors on both read and write calls.

**Comment by andreeaflorescu** [comment]
Since this looks like a corner case that wasn't handle, should we add a regression test here as well?

**Comment by gc-plp** [comment]
Yes, we want to have regression tests, but I was thinking that we could add them in a separate PR.

**Comment by andreeaflorescu** [comment]
Regression tests are typically merged in the same PR and the reviewers should check that the regression test fails before the fix and passes after the fix.

**Comment by gc-plp** [comment]
@andreeaflorescu, unit tests added, please review.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 3]
```diff
@@ -655,6 +655,17 @@ where
     }
 }
 
+#[cfg(test)]
+impl<S> VsockConnection<S>
```
Comment (iulianbarbu) [inline src/devices/src/virtio/vsock/csm/connection.rs:659]: Does it makes sense to add this impl inside the `mod tests` and then make the module `pub`?

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 4]
```diff
@@ -1339,4 +1339,33 @@ mod tests {
...
         assert_eq!(conn.fwd_cnt().0, 0);
     }
+
+    #[test]
+    fn test_regression_rxq_pop() {
+        let mut ctx = MuxerTestContext::new("regression_rxq_pop");
+        let peer_port = 1025;
+        let (mut stream, local_port) = ctx.local_connect(peer_port);
+
+        // Send some data
```
Comment (iulianbarbu) [inline src/devices/src/virtio/vsock/unix/muxer.rs:1349]: Nit: full stop at the end.

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 3 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs:659]: We could do that, but I think that this is the least invasive way of accessing private members declared in `connection.rs`

File: src/devices/src/virtio/vsock/unix/muxer.rs (hunk 4 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs]: Done

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 5]
```diff
@@ -1320,4 +1320,23 @@ mod tests {
 
         assert!(!ctx.muxer.has_pending_rx());
     }
+
+    #[test]
+    fn test_regression_handshake() {
+        let mut ctx = MuxerTestContext::new("regression_handshake");
+        let peer_port = 1025;
+
+        // Create a local connection
```
[the remark this answers is not in the document]
Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs]: Done

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 3 above)
Comment (iulianbarbu) [reply src/devices/src/virtio/vsock/csm/connection.rs:659]: I do not have a strong preference. My idea was to group all the tests relevant information in the `mod tests`, but I can not say that keeping this implementation outside the `mod tests` would counterfeit my idea of grouping. Eventually, we can group in the future such implementations for different structs in the `mod tests`, if they increase in number.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 6]
```diff
@@ -619,11 +650,27 @@ where
...
             .set_fwd_cnt(self.fwd_cnt.0)
     }
 }
 
+#[cfg(test)]
+impl<S> VsockConnection<S>
+where
+    S: Read + Write + AsRawFd,
+{
+    /// Get the fwd_cnt value from the connection
```
Comment (iulianbarbu) [inline src/devices/src/virtio/vsock/csm/connection.rs:663]: Nit: full stop as well.

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 3 above)
Comment (lauralt) [reply src/devices/src/virtio/vsock/csm/connection.rs:659]: > We could do that, but I think that this is the least invasive way of accessing private members declared in `connection.rs`

You can extend the struct implementation in `mod tests` without doing any other change and still be able to access private members; also the test module doesn't need to be `pub`. It's nice to keep the testing functionality in tests module because those functions won't be included in the `kcov` report, as opposed to how it is now implemented.

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 6 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: Done

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 3 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: Yep, you're right, that works too.

Comment (iulianbarbu) [reply src/devices/src/virtio/vsock/csm/connection.rs]: As far as I understood, this impl is used outside the mod tests, so the private members are not available there. That's why the methods must be declared as pub(crate) (combined with the fact that the struct is public).

Otherwise, if the impl would be included inside the mod tests, the module would need to be public in order to make use from the public interface exposed by the implementation outside the module/file scope.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 7]
```diff
@@ -749,6 +749,17 @@ mod tests {
         }
     }
 
+    #[cfg(test)]
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/vsock/csm/connection.rs:752]: You don't need to specify the `cfg(test)` macro because this implementation is defined in `mod tests` which is all only compiled for tests.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 8]
```diff
@@ -749,6 +749,17 @@ mod tests {
...
+    impl<S> VsockConnection<S>
+    where
+        S: Read + Write + AsRawFd,
+    {
+        /// Get the fwd_cnt value from the connection.
+        pub(crate) fn fwd_cnt(&mut self) -> Wrapping<u32> {
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/vsock/csm/connection.rs:758]: You shouldn't need `mut` here.

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 9]
```diff
@@ -1320,4 +1320,23 @@ mod tests {
 
         assert!(!ctx.muxer.has_pending_rx());
     }
+
+    #[test]
+    fn test_regression_handshake() {
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/vsock/unix/muxer.rs:1325]: What I noticed at another PR is that they also added a comment with the GitHub issue link that was related to the regression test. It would be nice to have it here as well.

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 10]
```diff
@@ -1339,4 +1339,33 @@ mod tests {
...
+        let conn = ctx.muxer.conn_map.get_mut(&key).unwrap();
+
+        // Forcefully insert another flag.
+        conn.insert_credit_update();
+
+        // Call recv twice in order to check that the connection is still
+        // in the rxq.
+        assert!(ctx.muxer.has_pending_rx());
+        ctx.recv();
+        assert!(ctx.muxer.has_pending_rx());
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/vsock/unix/muxer.rs:1368]: Shouldn't the assert be after the call to `recv`?

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 11]
```diff
@@ -1048,21 +1095,29 @@ mod tests {
 
         // Force a stale state, where the peer hasn't been updated on our credit situation.
         ctx.conn.last_fwd_cnt_to_peer = Wrapping(0);
-        ctx.conn.fwd_cnt = Wrapping(csm_defs::CONN_CREDIT_UPDATE_THRESHOLD as u32);
 
-        // Fake a data send from the peer, to bring us over the credit update threshold.
+        // Initialize the fwd counter at 6 bytes below the threshold
```
Comment (lauralt) [inline src/devices/src/virtio/vsock/csm/connection.rs:1099]: This comment is a bit misleading, as I would expect something like this: `initial_fwd_cnd = csm_defs::CONN_CREDIT_UPDATE_THRESHOLD as u32 - 6;` Can we update the comment or rename `CONN_CREDIT_UPDATE_THRESHOLD`?
Also, nit: full stop for all comments in this test.

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 12]
```diff
@@ -1339,4 +1339,33 @@ mod tests {
         // Check that fwd_cnt is 0 - "OK ..." was not accounted for.
         assert_eq!(conn.fwd_cnt().0, 0);
     }
+
+    #[test]
+    fn test_regression_rxq_pop() {
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/vsock/unix/muxer.rs:1344]: Same comment as for the other test. It would be useful to also have a link to the GitHub issue that this regression test is addressing.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 13]
```diff
@@ -718,6 +749,22 @@ mod tests {
         }
     }
 
+    #[cfg(test)]
```
Comment (lauralt) [inline src/devices/src/virtio/vsock/csm/connection.rs:752]: this can be removed.

File: src/devices/src/virtio/vsock/csm/connection.rs [hunk 14]
```diff
@@ -1048,21 +1095,29 @@ mod tests {
 
         // Force a stale state, where the peer hasn't been updated on our credit situation.
         ctx.conn.last_fwd_cnt_to_peer = Wrapping(0);
-        ctx.conn.fwd_cnt = Wrapping(csm_defs::CONN_CREDIT_UPDATE_THRESHOLD as u32);
 
-        // Fake a data send from the peer, to bring us over the credit update threshold.
+        // Initialize the fwd counter at 6 bytes below the threshold
+        let initial_fwd_cnd =
```
Comment (lauralt) [inline src/devices/src/virtio/vsock/csm/connection.rs:1100]:
```suggestion
        let initial_fwd_cnt =
```

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 11 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: Added a more comprehensive comment.

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 13 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: Removed

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 14 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: Typo fixed

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 7 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: Removed

File: src/devices/src/virtio/vsock/csm/connection.rs (hunk 8 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/csm/connection.rs]: `mut` removed

File: src/devices/src/virtio/vsock/unix/muxer.rs (hunk 9 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs:1325]: Link added

File: src/devices/src/virtio/vsock/unix/muxer.rs (hunk 12 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs:1347]: Link added

File: src/devices/src/virtio/vsock/unix/muxer.rs (hunk 10 above)
Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs:1375]: Actually, one assert was missing, because first of all we check that there is something pending, then we `recv`. After the first `recv` we should still have something pending, so we `recv` again.

Finally, we check that there is nothing pending.

File: src/devices/src/virtio/vsock/unix/muxer.rs [hunk 15]
```diff
@@ -1305,4 +1320,63 @@ mod tests {
 
         assert!(!ctx.muxer.has_pending_rx());
     }
+
+    #[test]
+    fn test_regression_handshake() {
+        // Address one of the issues found while fixing the following issue:
+        // https://github.com/firecracker-microvm/firecracker/pull/1910
```
Comment (lauralt) [inline src/devices/src/virtio/vsock/unix/muxer.rs:1327]: Hmm, actually this links to a PR, not issue; same for the next test.

Comment (gc-plp) [reply src/devices/src/virtio/vsock/unix/muxer.rs]: Pasted the wrong link :)

Fixed now.
